In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
import spacy

In [2]:
# Load dataset
dataset = pd.read_csv('dataset.csv')

# Drop rows with NaN values in 'clean_comment'
cleaned_dataset = dataset.dropna()

In [3]:
# Separate features and target
X_cleaned = cleaned_dataset['clean_comment']
y_cleaned = cleaned_dataset['category']

# Split the cleaned data into train and test sets (80-20 split)
X_train_cleaned, X_test_cleaned, y_train_cleaned, y_test_cleaned = train_test_split(X_cleaned, y_cleaned, test_size=0.2, random_state=42)


In [5]:
# Load spacy language model for POS tagging
nlp = spacy.load('en_core_web_sm')

In [6]:
# Function to extract custom features
def extract_custom_features(text):
    doc = nlp(text)
    word_list = [token.text for token in doc]

    # 1. Comment Length (number of characters)
    comment_length = len(text)

    # 2. Word Count
    word_count = len(word_list)

    # 3. Average Word Length
    avg_word_length = sum(len(word) for word in word_list) / word_count if word_count > 0 else 0

    # 4. Unique Word Count
    unique_word_count = len(set(word_list))

    # 5. Lexical Diversity
    lexical_diversity = unique_word_count / word_count if word_count > 0 else 0

    # 6. Count of POS Tags
    pos_count = len([token.pos_ for token in doc])

    # 7. Proportion of POS Tags
    pos_tags = [token.pos_ for token in doc]
    pos_proportion = {tag: pos_tags.count(tag) / word_count for tag in set(pos_tags)} if word_count > 0 else {}

    return {
        'comment_length': comment_length,
        'word_count': word_count,
        'avg_word_length': avg_word_length,
        'unique_word_count': unique_word_count,
        'lexical_diversity': lexical_diversity,
        'pos_count': pos_count,
        **pos_proportion  # Flattening the POS proportions
    }


In [7]:
# Apply the custom feature extraction
train_custom_features = pd.DataFrame([extract_custom_features(text) for text in X_train_cleaned])
test_custom_features = pd.DataFrame([extract_custom_features(text) for text in X_test_cleaned])

In [9]:
train_custom_features.head()

comment_length  word_count  avg_word_length  unique_word_count  \
0              51           7         6.428571                  7   
1              36           6         5.166667                  6   
2              64           9         6.222222                  9   
3             108          15         6.266667                 14   
4               6           1         6.000000                  1   

   lexical_diversity  pos_count      VERB      NOUN       ADV       AUX  ...  \
0           1.000000          7  0.428571  0.142857  0.142857  0.142857  ...   
1           1.000000          6  0.333333  0.333333       NaN       NaN  ...   
2           1.000000          9  0.222222  0.444444  0.111111       NaN  ...   
3           0.933333         15  0.200000  0.466667       NaN       NaN  ...   
4           1.000000          1       NaN  1.000000       NaN       NaN  ...   

        ADJ       NUM      PART  INTJ  CCONJ   X  SCONJ  DET  PUNCT  SYM  
0       NaN       NaN       NaN   NaN    NaN NaN    NaN  NaN    NaN  NaN  
1       NaN       NaN       NaN   NaN    NaN NaN    NaN  NaN    NaN  NaN  
2  0.222222       NaN       NaN   NaN    NaN NaN    NaN  NaN    NaN  NaN  
3  0.200000  0.066667  0.066667   NaN    NaN NaN    NaN  NaN    NaN  NaN  
4       NaN       NaN       NaN   NaN    NaN NaN    NaN  NaN    NaN  NaN  

[5 rows x 23 columns]

In [10]:
# Replace NaN values in POS tag proportions with 0
train_custom_features.fillna(0, inplace=True)
test_custom_features.fillna(0, inplace=True)

In [11]:
test_custom_features.isnull().sum()

comment_length       0
word_count           0
avg_word_length      0
unique_word_count    0
lexical_diversity    0
pos_count            0
VERB                 0
NOUN                 0
ADJ                  0
PRON                 0
ADP                  0
PROPN                0
AUX                  0
ADV                  0
NUM                  0
CCONJ                0
INTJ                 0
DET                  0
PART                 0
SCONJ                0
X                    0
PUNCT                0
SYM                  0
dtype: int64

In [12]:
# Apply TfidfVectorizer with trigram setting and max_features=10000
tfidf = TfidfVectorizer(ngram_range=(1, 3), max_features=10_000)
X_train_tfidf = tfidf.fit_transform(X_train_cleaned)
X_test_tfidf = tfidf.transform(X_test_cleaned)

In [13]:
# Convert TF-IDF to DataFrame
X_train_tfidf_df = pd.DataFrame(X_train_tfidf.toarray(), columns=tfidf.get_feature_names_out())
X_test_tfidf_df = pd.DataFrame(X_test_tfidf.toarray(), columns=tfidf.get_feature_names_out())

In [14]:
# Combine TF-IDF and custom features
X_train_combined = pd.concat([X_train_tfidf_df.reset_index(drop=True), train_custom_features.reset_index(drop=True)], axis=1)
X_test_combined = pd.concat([X_test_tfidf_df.reset_index(drop=True), test_custom_features.reset_index(drop=True)], axis=1)

In [15]:
X_train_combined

000  000 000  000 crore  000 rupee  100  100 crore  100 time  100 year  \
0      0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
1      0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
2      0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
3      0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
4      0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
...    ...      ...        ...        ...  ...        ...       ...       ...   
29324  0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
29325  0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
29326  0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
29327  0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   
29328  0.0      0.0        0.0        0.0  0.0        0.0       0.0       0.0   

       1000  1000 note  ...       ADJ       NUM      PART      INTJ  CCONJ  \
0       0.0        0.0  ...  0.000000  0.000000  0.000000  0.000000    0.0   
1       0.0        0.0  ...  0.000000  0.000000  0.000000  0.000000    0.0   
2       0.0        0.0  ...  0.222222  0.000000  0.000000  0.000000    0.0   
3       0.0        0.0  ...  0.200000  0.066667  0.066667  0.000000    0.0   
4       0.0        0.0  ...  0.000000  0.000000  0.000000  0.000000    0.0   
...     ...        ...  ...       ...       ...       ...       ...    ...   
29324   0.0        0.0  ...  0.250000  0.000000  0.000000  0.000000    0.0   
29325   0.0        0.0  ...  0.100000  0.000000  0.000000  0.000000    0.0   
29326   0.0        0.0  ...  0.000000  0.000000  0.000000  0.000000    0.0   
29327   0.0        0.0  ...  0.189189  0.000000  0.000000  0.054054    0.0   
29328   0.0        0.0  ...  0.000000  0.000000  0.000000  0.000000    0.0   

         X  SCONJ  DET  PUNCT  SYM  
0      0.0    0.0  0.0    0.0  0.0  
1      0.0    0.0  0.0    0.0  0.0  
2      0.0    0.0  0.0    0.0  0.0  
3      0.0    0.0  0.0    0.0  0.0  
4      0.0    0.0  0.0    0.0  0.0  
...    ...    ...  ...    ...  ...  
29324  0.0    0.0  0.0    0.0  0.0  
29325  0.0    0.0  0.0    0.0  0.0  
29326  0.0    0.0  0.0    0.0  0.0  
29327  0.0    0.0  0.0    0.0  0.0  
29328  0.0    0.0  0.0    0.0  0.0  

[29329 rows x 10023 columns]

In [16]:
import lightgbm as lgb

In [17]:
model = lgb.LGBMClassifier(
    objective="multiclass",
    num_class=3,
    metric="multi_logloss",
    is_unbalance=True,
    class_weight="balanced",
    reg_alpha=0.1,  # L1 regularization
    reg_lambda=0.1,  # L2 regularization
    learning_rate=0.07940802964603502,
    n_estimators=496,
    max_depth=20,
)

In [18]:
# Fit the model on the resampled training data
model.fit(X_train_combined, y_train_cleaned)

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.284654 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 136998
[LightGBM] [Info] Number of data points in the train set: 29329, number of used features: 4461
[LightGBM] [Info] Start training from score -1.098612
[LightGBM] [Info] Start training from score -1.098612
[LightGBM] [Info] Start training from score -1.098612
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [War

LGBMClassifier(class_weight='balanced', is_unbalance=True,
               learning_rate=0.07940802964603502, max_depth=20,
               metric='multi_logloss', n_estimators=496, num_class=3,
               objective='multiclass', reg_alpha=0.1, reg_lambda=0.1)

In [19]:
# Predict on the test set
from sklearn.metrics import accuracy_score
y_pred = model.predict(X_test_combined)
accuracy = accuracy_score(y_test_cleaned, y_pred)
accuracy

0.8561298240829128

In [20]:
from sklearn.metrics import classification_report
# Generate classification report
report = classification_report(y_test_cleaned, y_pred)
print(report)

              precision    recall  f1-score   support

          -1       0.79      0.76      0.77      1647
           0       0.84      0.96      0.90      2510
           1       0.91      0.82      0.86      3176

    accuracy                           0.86      7333
   macro avg       0.85      0.85      0.84      7333
weighted avg       0.86      0.86      0.85      7333



### 
              precision    recall  f1-score   support

          -1       0.81      0.77      0.79      1647
           0       0.85      0.97      0.90      2510
           1       0.92      0.83      0.87      3176

    accuracy                           0.87      7333
   macro avg       0.86      0.86      0.85      7333
weighted avg       0.87      0.87      0.86      7333
`